rpc: add additional ban time fields to listbanned

jarolrod opened the pull request:
This PR adds a `ban_duration` and `time_remaining` field to the `listbanned` RPC command. Thanks to jonatack, this PR also expands the `listbanned` test coverage to include these new fields

It's useful to keep track of `ban_duration` as this is another data point on which to sort banned peers. I found this helpful in adding additional context columns to the GUI `bantablemodel` as part of a follow-up PR. As [suggested](https://github.com/bitcoin/bitcoin/pull/21602#issuecomment-813486134) by jonatack, `time_remaining` is another useful user-centric data point.

Since a ban always expires after its created, the `ban_created` field is now placed before the `banned_until` field. This new ordering is more logical.

This PR also improves the `help listbanned` output by providing additional context to the descriptions of the `address`, `ban_created`, and `banned_until` fields.



**Master: listbanned**
```
[
  {
    "address": "1.2.3.4/32",
    "banned_until": 1617691101,
    "ban_created": 1617604701
  },
  {
    "address": "135.181.41.129/32",
    "banned_until": 1649140716,
    "ban_created": 1617604716
  }
]
```

**PR: listbanned**
```
[
  {
    "address": "1.2.3.4/32",
    "ban_created": 1617775773,
    "banned_until": 1617862173,
    "ban_duration": 86400,
    "time_remaining": 86392
  },
  {
    "address": "3.114.211.172/32",
    "ban_created": 1617753165,
    "banned_until": 1618357965,
    "ban_duration": 604800,
    "time_remaining": 582184
  }
]
```

jonatack commented:
Would this cause versioning/compatibility issues with previous/existing ban entries? (test coverage needed). It may be better to avoid changing `CBanEntry`, and have the RPC handle the calculation.

As a human user, I've found myself wishing for a "time_remaining" field in `listbanned`.

jonatack commented on src/rpc/net.cpp:
```diff
@@ -749,9 +749,10 @@ static RPCHelpMan listbanned()
+                        {RPCResult::Type::STR, "address", "The IP/Subnet of the banned node"},
+                        {RPCResult::Type::NUM_TIME, "banned_until", "The " + UNIX_EPOCH_TIME + " the node is banned until"},
+                        {RPCResult::Type::NUM_TIME, "ban_created", "The " + UNIX_EPOCH_TIME + " the ban was created"},
+                        {RPCResult::Type::NUM_TIME, "ban_time", "The amount of time the ban will last for in seconds"},
```
```suggestion
                        {RPCResult::Type::NUM_TIME, "ban_time", "The ban duration, in seconds"},
```

jarolrod replied:
Thanks for making my original text look like a rube goldberg machine 🤗

jarolrod replied:
addressed in 8e34b5ed06f3ca0865b9b7dca6e31f2e95c6f5c8

jarolrod commented:
@jonatack 
> It may be better to avoid changing `CBanEntry`, and have the RPC handle the calculation.

You're right, taking into consideration compatibility guarantees, performing the calculation is the simplest path.
 I will revise this PR to perform the calculation and avoid changing `CBanEntry` and `Banman`. 

> As a human user, I've found myself wishing for a "time_remaining" field in `listbanned`.

I will look into adding this field, thanks for the review!

jarolrod commented:
updated from 9c896134b12f012d0736742693df4b50c133e506 to 8e34b5ed06f3ca0865b9b7dca6e31f2e95c6f5c8, addressed @jonatack suggestions:
- perform the calculation inside the RPC command instead of changing `CBanEntry` and `Banman`. This avoids any compatibility issues.
- Add additional `ban_time_remaining` field to `listbanned`

sipa commented on src/rpc/net.cpp:
```diff
@@ -777,6 +778,7 @@ static RPCHelpMan listbanned()
         rec.pushKV("banned_until", banEntry.nBanUntil);
         rec.pushKV("ban_created", banEntry.nCreateTime);
         rec.pushKV("ban_time", (banEntry.nBanUntil - banEntry.nCreateTime));
+        rec.pushKV("ban_time_remaining", (banEntry.nBanUntil - GetTime()));
```
Maybe query the time once before the loop, and use it for every iteration? This is needlessly slow and inconsistent.

jarolrod replied:
whoops, brain fart, forgot we were in a loop

jarolrod replied:
addressed in 833bebdece0f6fac5fecf602e88ef7c114f2416c

jarolrod commented:
updated from 8e34b5e to 833bebdece0f6fac5fecf602e88ef7c114f2416c, addressed @sipa suggestion:
- GetTime() was being called inside of a loop due to an oversight. This is unnecessary and can produce inconsistent results. This moves GetTime() out of the loop so it is only called once.

sipa commented:
utACK 833bebdece0f6fac5fecf602e88ef7c114f2416c

jonatack commented on src/rpc/net.cpp:
```diff
@@ -776,6 +780,8 @@ static RPCHelpMan listbanned()
         rec.pushKV("banned_until", banEntry.nBanUntil);
         rec.pushKV("ban_created", banEntry.nCreateTime);
+        rec.pushKV("ban_time", (banEntry.nBanUntil - banEntry.nCreateTime));
+        rec.pushKV("ban_time_remaining", (banEntry.nBanUntil - current_time));
```
suggestions
```diff
@@ -751,10 +751,10 @@ static RPCHelpMan listbanned()
                 {RPCResult::Type::OBJ, "", "",
                     {
                         {RPCResult::Type::STR, "address", "The IP/Subnet of the banned node"},
-                        {RPCResult::Type::NUM_TIME, "banned_until", "The " + UNIX_EPOCH_TIME + " the node is banned until"},
                         {RPCResult::Type::NUM_TIME, "ban_created", "The " + UNIX_EPOCH_TIME + " the ban was created"},
-                        {RPCResult::Type::NUM_TIME, "ban_time", "The ban duration, in seconds"},
-                        {RPCResult::Type::NUM_TIME, "ban_time_remaining", "The amount of seconds until the ban expires"},
+                        {RPCResult::Type::NUM_TIME, "banned_until", "The " + UNIX_EPOCH_TIME + " the ban expires"},
+                        {RPCResult::Type::NUM_TIME, "ban_duration", "The ban duration, in seconds"},
+                        {RPCResult::Type::NUM_TIME, "time_remaining", "The time remaining until the ban expires, in seconds"},
                     }},
             }},
                 RPCExamples{
@@ -770,7 +770,7 @@ static RPCHelpMan listbanned()
 
     banmap_t banMap;
     node.banman->GetBanned(banMap);
-    int64_t current_time = GetTime();
+    const int64_t current_time{GetTime()};
 
     UniValue bannedAddresses(UniValue::VARR);
     for (const auto& entry : banMap)
@@ -778,10 +778,10 @@ static RPCHelpMan listbanned()
         const CBanEntry& banEntry = entry.second;
         UniValue rec(UniValue::VOBJ);
         rec.pushKV("address", entry.first.ToString());
-        rec.pushKV("banned_until", banEntry.nBanUntil);
         rec.pushKV("ban_created", banEntry.nCreateTime);
-        rec.pushKV("ban_time", (banEntry.nBanUntil - banEntry.nCreateTime));
-        rec.pushKV("ban_time_remaining", (banEntry.nBanUntil - current_time));
+        rec.pushKV("banned_until", banEntry.nBanUntil);
+        rec.pushKV("ban_duration", (banEntry.nBanUntil - banEntry.nCreateTime));
+        rec.pushKV("time_remaining", (banEntry.nBanUntil - current_time));
 
         bannedAddresses.push_back(rec);
     }
```

result
```
$ bitcoin-cli -signet help listbanned
listbanned

List all manually banned IPs/Subnets.

Result:
[                              (json array)
  {                            (json object)
    "address" : "str",         (string) The IP/Subnet of the banned node
    "ban_created" : xxx,       (numeric) The UNIX epoch time the ban was created
    "banned_until" : xxx,      (numeric) The UNIX epoch time the ban expires
    "ban_duration" : xxx,      (numeric) The ban duration, in seconds
    "time_remaining" : xxx     (numeric) The time remaining until the ban expires, in seconds
  },

$ bitcoin-cli -signet setban 51.15.12.61 add 30

$ bitcoin-cli -signet listbanned
[
  {
    "address": "51.15.12.61/32",
    "ban_created": 1617727507,
    "banned_until": 1617727543,
    "ban_duration": 36,
    "time_remaining": 29
  }
]
```

jarolrod replied:
Thanks for the in-depth review! Special thanks for contributing a test. Addressed in c5c920f

jonatack reviewed (commented):
Tested approach ACK

Various ideas and suggestions below grouped into one comment. If you retouch, perhaps update the relevant help in each commit that adds a field, so each commit is self-contained. Also, needs a release note.

jonatack committed:
test: increase listbanned unit test coverage

Add test coverage for the new ban_duration and time_remaining fields.
While here, some code improvements.

jonatack commented:
@jarolrod feel free to pull in the last commit in this branch https://github.com/jonatack/bitcoin/commits/pr21602-suggestions to add unit test coverage for these changes

jarolrod committed:
rpc: swap position of banned_until and ban_created fields

A ban expires after its creation. Therefore, for the listbanned RPC,
position banned_until after ban_created in help and output.

jarolrod committed:
doc: improve listbanned help

Add descriptions for the address, ban_created, and banned_until fields.

jarolrod committed:
rpc: add ban_duration field to listbanned

jarolrod committed:
rpc: add time_remaining field to listbanned

jarolrod committed:
doc: release notes for new listbanned fields

jarolrod commented:
updated from 833bebd to c5c920f, addressed @jonatack [suggestions](https://github.com/bitcoin/bitcoin/pull/21602#discussion_r608016518)

- swap positions of `ban_created` and `banned_until`
- rename `ban_time` to `ban_duration`
- rename `ban_time_remaining` to `time_remaining`
- drop  `#include util/time.h`, this is already included by something else. 
  - Other commands already use `time.h` functions in this part of the code.
- Add test coverage for new `listbanned` fields, thanks to jon
- Add release notes for new fields + `banned_until` reposition

jonatack commented on doc/release-notes.md:
```diff
@@ -104,6 +104,10 @@ Updated RPCs
 
+- The `listbanned` RPC now returns two new numeric fields: `ban_duration` and `time_remaining`.
+  Respectively, these new fields indicate the duration of a ban and the time remaining until a ban expires,
+  both in seconds. Additionally, the `ban_created` field is repositioned to come before `banned_until`.
```
```suggestion
  both in seconds. Additionally, the `ban_created` field is repositioned to come before `banned_until`.  (#21602)
```

jarolrod replied:
done in d3b0b08b0f04d2f1dbebbafd7ab0384dfe045dec

jonatack reviewed (commented):
Tested ACK c5c920f394891f1d6958913330da564bf24a71a9

jonatack commented:
re-ACK d3b0b08b0f04d2f1dbebbafd7ab0384dfe045dec

There's a small potential race in the added unit tests. I didn't hit while running the test, even with valgrind to slow it down, but maybe something like this would provide a margin:
```diff
@@ -277,6 +277,7 @@ BOOST_AUTO_TEST_CASE(rpc_ban)
 
     BOOST_CHECK_NO_THROW(r = CallRPC(std::string("setban 127.0.0.0/24 add 200")));
     BOOST_CHECK_NO_THROW(r = CallRPC(std::string("listbanned")));
+    const int64_t now{GetTime()};
     ar = r.get_array();
     o1 = ar[0].get_obj();
     adr = find_value(o1, "address");
@@ -284,12 +285,13 @@ BOOST_AUTO_TEST_CASE(rpc_ban)
     const int64_t ban_created{find_value(o1, "ban_created").get_int64()};
     const int64_t ban_duration{find_value(o1, "ban_duration").get_int64()};
     const int64_t time_remaining{find_value(o1, "time_remaining").get_int64()};
-    const int64_t now{GetTime()};
     BOOST_CHECK_EQUAL(adr.get_str(), "127.0.0.0/24");
     BOOST_CHECK(banned_until > now);
     BOOST_CHECK(banned_until - now <= 200);
-    BOOST_CHECK_EQUAL(ban_duration, banned_until - ban_created);
-    BOOST_CHECK_EQUAL(time_remaining, banned_until - now);
+    BOOST_CHECK(ban_duration >= banned_until - ban_created - 1);
+    BOOST_CHECK(ban_duration <= banned_until - ban_created + 1);
+    BOOST_CHECK(time_remaining >= banned_until - now - 1);
+    BOOST_CHECK(time_remaining <= banned_until - now + 1);
```

promag reviewed (commented):
Code review d3b0b08b0f04d2f1dbebbafd7ab0384dfe045dec

I'd drop dd3c8eaa3399b28dc78a883ff78cbe7cc5c31b5b, but if kept then mark it move-only or squash with c95c61657afd058b46549fb3d65633d7c736f5fc.

I'd also squash 5456b345312857981cb426712f0665800c682e09 with 3e978d1a5dbd43f85bd03e759984ab1f209d6e34.

hebasto commented:
Concept ACK on:
> I found this helpful in adding additional context columns to the GUI `bantablemodel` as part of a follow-up PR.

hebasto commented on src/rpc/net.cpp:
```diff
@@ -767,15 +769,18 @@ static RPCHelpMan listbanned()
 
     banmap_t banMap;
     node.banman->GetBanned(banMap);
+    const int64_t current_time{GetTime()};
```
https://github.com/bitcoin/bitcoin/blob/dd01dc6a1d23e1f316940cd1092ad716216331aa/src/util/time.h#L40-L44

```suggestion
    const int64_t current_time{GetSystemTimeInSeconds()};
```

MarcoFalke replied:
banman uses `GetTime()` (mockable), so mixing non-mockable and mockable should be avoided

hebasto replied:
Ah, thanks!

hebasto replied:
So it should be `GetTime<std::chrono::seconds>`?

MarcoFalke replied:
I guess so, but that refactor should also be done for nCreateTime, banman, ...

hebasto reviewed (commented):
Approach ACK d3b0b08b0f04d2f1dbebbafd7ab0384dfe045dec.

> I found this helpful in adding additional context columns to the GUI `bantablemodel` as part of a follow-up PR.

It seems, that adding additional time-related fields to the "Banned peers" table in the GUI does not depend on this PR :)

hebasto reviewed (approved):
ACK d3b0b08b0f04d2f1dbebbafd7ab0384dfe045dec, tested on Linux Mint 20.1 (x86_64).

maflcko reviewed (commented):
review ACK d3b0b08b0f04d2f1dbebbafd7ab0384dfe045dec 🕙

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK d3b0b08b0f04d2f1dbebbafd7ab0384dfe045dec 🕙
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgCnQv+Pt8SJP4wssUe0R41DmjzJ0d0nEqEiWA38vaipjeD6zNgy5yBhuOmtsm/
vdlSvsOogPjoXM7nrCEaODFRdM8Sj9cfFtPi94U6dmU8lIv92urpREeUm92Zf1PI
2z0mMJwktzszLVK1t3p6Qh90iV3RDy/OE/kkKYOjs41wLMtt1akHchkGVVMuA6P9
w3L8rVoG/6dlWsWd9H9bWvJ/0QEm12zK9hlHZONGBCHC4NmM/QEYLFqULmsAVmuw
7h9XysqcQVCVTFcUD4Qk7y2WOM3vtW3OI2N+DzLK9tOcYvJ0LlGByNe5RSGnLzgl
wewt7k2piUr8qSQNKIvCSIFoy3K3BuVAGWpvFmMgR6tEidhrMhMxFc+rCcvPQbDI
gHypLP4J3DwU55PQ/vAO4IpsSudiVmgUbO9gkrZCJ5r9qMTgJhWtN7epkTgcVLUn
aXbS1eLlMrEcnRt1VdLc42VOo4P755/lNg/QmZcQTub3Z46KxhOhf+gSTOut3KL6
3lJzmjRC
=qw8I
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `4d1380dd6579b539625e1fcb8f85f0b5ec4b795d7be43d7e91a24d790b488c20  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401084d1380dd6579b539625e1fcb8f85f0b5ec4b795d7be43d7e91a24d790b488c20f0101462770d124841b01ff0ce7793b1acf408fff0101c56871ec362f1d380509611dc19cf3b08f020dcdc6e2fc8feff1f12a06692917180f36a768de2b142508a03957642582b692008f020a55bafad32cfabc661015a185de2a9f7014406a99283e2df29753bbb714f98b308f1046072d119f008ebc5506bdc1f98240083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010764ee9ddfca05467e2b466c102bb717108f020c7c3714368de3fd69b6aec9508b64eb77bb58fbe724927e3060e264987f8043308f02019227d6956202123fff612352bbfc942b05c19d3bbefb133adb0573fdf25a23708f1046072def2f008e0f296a5fff4864a0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a00bd7679cc17b31e85365d6b6e8307908f1046072def3f00873a27a53d94974410083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d255f9477bbc1db158076040870899d808f1046072def2f00834313087b66151f20083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented:
(edited OP to remove `@`)

MarcoFalke commented on src/test/rpc_tests.cpp:
```diff
@@ -280,11 +280,16 @@ BOOST_AUTO_TEST_CASE(rpc_ban)
+    BOOST_CHECK(banned_until > now);
+    BOOST_CHECK(banned_until - now <= 200);
+    BOOST_CHECK_EQUAL(ban_duration, banned_until - ban_created);
+    BOOST_CHECK_EQUAL(time_remaining, banned_until - now);
```
the may fail:

```
Running 1 test case...
test/rpc_tests.cpp(293): error: in "rpc_tests/rpc_ban": check time_remaining == banned_until - now has failed [200 != 199]

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
```

diff:

```diff
diff --git a/src/test/rpc_tests.cpp b/src/test/rpc_tests.cpp
index 3b6faf7bbb..be0d79aa86 100644
--- a/src/test/rpc_tests.cpp
+++ b/src/test/rpc_tests.cpp
@@ -277,6 +277,7 @@ BOOST_AUTO_TEST_CASE(rpc_ban)
 
     BOOST_CHECK_NO_THROW(r = CallRPC(std::string("setban 127.0.0.0/24 add 200")));
     BOOST_CHECK_NO_THROW(r = CallRPC(std::string("listbanned")));
+    std::this_thread::sleep_for(std::chrono::milliseconds(1100)); // CPU scheduler puts this thread to sleep
     ar = r.get_array();
     o1 = ar[0].get_obj();
     adr = find_value(o1, "address");
```

Can be fixed by using mocktime (SetTime()), or by relaxing the test.

Also, the whole test can be moved to the functional test suite. For example, `./test/functional/rpc_setban.py`

MarcoFalke replied:
See #21676
